### ISSUE TITLE
Setting continuous integration

### ISSUE THREAD

**Issue by amirouche** [udem-dlteam/libs#4]
Can we setup continuous integration at, say Cirrus CI, since it seems like the most advanced CI system.

ref: https://mailman.iro.umontreal.ca/pipermail/gambit-list/2019-November/009222.html

**Comment by lassik** [comment]
@weinholt knows a lot about CI and says Circle and GitLab have more advanced features than Cirrus. But Cirrus has a lot of platforms (Linux/Docker, Windows, Mac, FreeBSD) and is fast and easy. We could start with it. All of these new CI hosts ought to be easy to set up, so if it doesn't work out, it's not hard to change.

**Comment by lassik** [comment]
I love the ultra-simple UI of Cirrus: https://cirrus-ci.com/github/weinholt/ It's very fast and has only the information you need.

**Comment by amirouche** [comment]
I can make a pull request if you want

**Comment by lassik** [comment]
That would be great, but let's wait for @feeley to confirm. He probably has to register a Cirrus account and then enable it for the udem-dlteam organization from the GitHub settings.

One thing that would be good to investigate right away, is how publishing of build artifacts works. It would be great to always be able to download gsc and gsi binaries of the latest successful CI build.

**Comment by amirouche** [comment]
Le lun. 4 nov. 2019 à 09:38, Lassi Kortela <notifications@github.com> a
écrit :

> One thing that would be good to investigate right away, is how publishing
> of build artifacts works. It would be great to always be able to download
> gsc and gsi binaries of the latest successful CI build.
>

This is the interesting documentation:
https://cirrus-ci.org/guide/writing-tasks/#artifacts-instruction

This will require to setup cirrus at gambit/gambit. Maybe it is easier for
the time being to build gambit at each CI run?

**Comment by amirouche** [comment]
Related to code coverage: https://github.com/gambit/gambit/issues/94

**Comment by lassik** [comment]
We could also try to set it up for udem-dlteam/gambit first

**Comment by lassik** [comment]
Just realized I already have a Cirrus account, so I can add Cirrus to my personal fork of Gambit and find out how it goes. That's the fastest way to find out. If it's good we can then add it to gambit/gambit and this udem-dlteam. If there are problems, we save some account creation hassles and can try another provider.

**Comment by feeley** [comment]
I also realized I have one and just configured it for `udem-dlteam/gambit` to see how it goes.

**Comment by lassik** [comment]
CI often takes many tries to get right. Since CI services build all branches of the repo, it's usually easiest to do the build experimentation in a topic branch and merge to master once it's working.

**Comment by lassik** [comment]
BTW how much are the gsc and gsi executables affected by the `./configure --prefix=...` argument? If the CI builds them with a particular prefix, can those binaries be copied onto another computer and still be usable from an arbitrary directory?

**Comment by lassik** [comment]
> CI agent stopped responding!
> Container errored with 'OOMKilled'

Out of memory, and we're not even building with `--enable-single-host` :|

**Comment by feeley** [comment]
I have created a basic `.cirrus.yml` file and is currently executing. For future reference: it is important to pull the tags from the repo first so that the build process knows we need to bootstrap the compiler first instead of immediately compiling the (stale) .c files.

I'm not sure why it is giving a `Container errored with 'OOMKilled'` error.  Maybe the `-j` on the make is too much...

@amirouche @lassik Since you now have write access to both repositories, I will let you continue the experimenting with Cirrus CI.  If it looks good then we can move the changes to `gambit/gambit`.

We need to verify that Cirrus CI allows checking the main platforms (linux, macOS and Windows), and with the different build settings that are currently done with Travis and Appveyor.

Concerning the `--prefix=...`, the build process burns the path into the executable as the default "Gambit installation directory".  If you want to use the executable on another machine, you will need to move the `bin`, `lib` and `include` directories, and use an explicit `-:~~=<path>` runtime option to inform the runtime system where those directories are.  Here is some relevant information from `INSTALL.txt`:

```
If the --prefix option is not used, the default is to install all
files in /usr/local/Gambit and its subdirectories, namely "bin",
"lib", "include", etc.  The files that would normally go in these
subdirectories can be redirected to other directories using
the following configure options:

  --bindir=DIR            executables (gsi, gsc, ...)
  --libdir=DIR            libraries (libgambit.a, syntax-case.scm, ...)
  --includedir=DIR        C header files (gambit.h, ...)
  --docdir=DIR            documentation (gambit.pdf, gambit.html, ...)
  --infodir=DIR           info documentation (gambit.info, ...)
  --datadir=DIR           read-only architecture-independent data (gambit.el)

Note that the install target of the makefiles supports the DESTDIR
environment variable which allows staging an install to a specific
directory.  The command:

  % make install DESTDIR=/Users/feeley/stage

will install all the files in /Users/feeley/stage as though it was the
root of the filesystem.
```

**Comment by lassik** [comment]
Thanks! I'm quite excited about this so I can get to it right away. I'll create a `lassik/cirrus` branch in `udem-dlteam/gambit` for experimenting (IMHO it's clearest if each topic branch name says who is working on it, otherwise old mystery branches tend to be left behind in repos).

Not sure how to deal with the prefix for CI builds. Maybe something like `./configure --prefix=/usr/local/gambit/$CIRRUS_BRANCH`?

**Comment by feeley** [comment]
Does it matter? Is the filesystem persistent?

For the OOM error, try removing the `-j` on the make.

**Comment by lassik** [comment]
It doesn't matter during the build as the CI file system vanishes afterwards. But we can pick some files to save from each successful build as artifacts so people can download them. It would probably be useful to save the gsc and gsi binaries in this way. If those depend on the prefix, then that prefix would have to be something that is likely to work sensibly on the personal computers of people who download the binaries.

**Comment by feeley** [comment]
Ok I understand.  I guess it could be the commit hash.  I guess the `$CIRRUS_BRANCH` is related to that.  Note that historically I have used `/usr/local/Gambit` (capital G).

**Comment by lassik** [comment]
If `/usr/local/Gambit` is already in use we could stick with it. But in case someone has installed a stable release of Gambit in there, we'd do well to think of some prefix that doesn't overwrite that one with a cutting-edge version. The git branch or hash would be good since one can then try multiple experimental branches on the same computer. Is the hash too precise? It would change with every new commit to the same branch. The git tag is probably not precise enough, as bleeding-edge builds are not tagged yet.

**Comment by lassik** [comment]
`$CIRRUS_BRANCH` is just the Git branch. It sets other similar envars also: https://cirrus-ci.org/guide/writing-tasks/#environment-variables

**Comment by feeley** [comment]
Here's a useful configure option:

```
  --enable-multiple-versions
                          multiple installed versions are supported
```

It will put the installation in `$prefix/<version>` where version is something like `v4.9.3` and will create the symbolic link `current` to point to the latest installed version.  So using the commit hash as a "version" will not interfere.

**Comment by lassik** [comment]
Very nice! How is the `<version>` set and how does it determine which version is latest?

**Comment by lassik** [comment]
Without `-j` the `make` succeeds in Cirrus. But one unit test fails:

```
[223|  0]  94% ###############. 110.9s 13-modules/prim_port.scm
*** FAILED 13-modules/prim_port.scm WITH EXIT CODE HI=1 LO=0
(call-with-output-process "echo" ##newline) in (include "~~lib/_prim-port#.scm") returned #!void but expected #<os-exception #2>
(with-output-to-process "echo" ##newline) in (include "~~lib/_prim-port#.scm") returned #<os-exception #3> but expected #<os-exception #4>
(call-with-output-process "echo" ##newline) in (namespace ("##")) returned #<os-exception #5> but expected #<os-exception #2>
(with-output-to-process "echo" ##newline) in (namespace ("##")) returned #<os-exception #6> but expected #<os-exception #4>
```

**Comment by feeley** [comment]
Hey the tests passed except for 1 out of 237.  Don't know what's happening with "echo".

I must say the build machine is underpowered... it took 110 seconds for the tests, and on my 2013 laptop it takes ~10 seconds for a `make ut`.

**Comment by feeley** [comment]
For the failing unit test, the result of `(call-with-output-process "echo" ##newline)` is compared to the result of `(##call-with-output-process "echo" ##newline)`. It seems the first one is returning an OS exception and the second one is executing fine without an exception.  This is very strange as these should be equivalent.

Note that `echo` is used because it is a program (or command) that exists on unix and windows, in this particular case a newline is sent to the echo program that ignores it, so nothing bad should happen.  It has never caused problems in the past on my development machines, and Travis.  Could it be a transient problem? Could it be related to how slow the machine is? Maybe the build should be repeated to make sure...

**Comment by feeley** [comment]
Hmmm I misread the errors.  It seems they all (except for one) fail with an os-exception... Could it be that there's a limit on subprocesses?

**Comment by lassik** [comment]
> I must say the build machine is underpowered... it took 110 seconds for the tests, and on my 2013 laptop it takes ~10 seconds for a `make ut`.

Their test runners are rented on-demand from Google cloud. Since we are using the free service, they probably use some of the slower offerings. It is possible to plug your own build server into Cirrus for free (and to some other CI services, e.g. GitLab). I'm not familiar with the technical details. The paid accounts on these services are presumably faster/have more RAM.

Since some of the tests do I/O, there may be some additional factor causing slowdown - not sure. 10x slowdown is quite a lot.

The FreeBSD jobs also often take a lot of time to get going because there's limited capacity and a queue for the free service. The others don't have FreeBSD at all.

> For the failing unit test, the result of `(call-with-output-process "echo" ##newline)` is compared to the result of `(##call-with-output-process "echo" ##newline)`. It seems the first one is returning an OS exception and the second one is executing fine without an exception. This is very strange as these should be equivalent.

> Note that `echo` is used because it is a program (or command) that exists on unix and windows, in this particular case a newline is sent to the echo program that ignores it, so nothing bad should happen. It has never caused problems in the past on my development machines, and Travis. Could it be a transient problem? Could it be related to how slow the machine is? Maybe the build should be repeated to make sure...

Would be quite strange if speed problems cause an exception. I'll add the artifact settings which triggers another build. If the echo still fails, we could add that a debug print to get details about the exception.

**Comment by feeley** [comment]
> Very nice! How is the `<version>` set and how does it determine which version is latest?

The version is `$PACKAGE_VERSION` from the configure script.  My point is that if the user's Gambit was configures with the `--enable-multiple-versions` there will be no conflict if you use `--prefix=/usr/local/Gambit/<commit_hash>`.

Perhaps the configure script should have a mechanism to specify the version, other than using `$PACKAGE_VERSION` which would be the default.  Maybe a `--enable-version=<commit_hash>`.

**Comment by lassik** [comment]
Would be nice to have an `--enable-single-host` build as well, but it will probably be extremely slow and possibly run out of RAM. We get what we pay for :)

**Comment by lassik** [comment]
https://cirrus-ci.org/guide/linux/ says:

> Linux Community Cluster is a Kubernetes cluster running on Google Kubernetes Engine that is available free of charge for Open Source community. Paying customers can also use Community Cluster for personal private repositories or buy CPU time with compute credits for their private organization repositories.
> 
> Community Cluster is configured the same way as anyone can configure a personal GKE cluster as described here.
> 
> By default a container is given 2 CPUs and 4 GB of memory but it can be configured in .cirrus.yml:
>
> Containers on Community Cluster can use maximum 8.0 CPUs and up to 24 GB of memory.

**Comment by feeley** [comment]
I like the idea of running the CI locally.  But we can explore that later if Cirrus gives the features we need.

You should try `--enable-single-host` in the next round.  As long as there's no parallel build I'm confident it will pass.

**Comment by lassik** [comment]
Can you estimate how many gigs of RAM it needs?

**Comment by lassik** [comment]
We can try being greedy and using the maximum of 8 CPUs and 24 GB RAM. Maybe they have a fixed capacity, so if you request higher specs you'll have to wait longer in the queue for the build to start.

**Comment by feeley** [comment]
> Can you estimate how many gigs of RAM it needs?

The memory usage of gcc is probably the limiting factor and it greatly depends on the gcc version. For gcc 8 which I use regularly, I haven't seen more that 1 GB RSS per gcc compilation.

It should be viewed as a bug if 4 GB RAM is not enough to build Gambit (for a non-parallel build).

**Comment by lassik** [comment]
Good to know. I tried using max CPU and RAM and `make -j` - now it went really fast. The log looks like it tries to run configure and compilation thrice (???) and finally fails at something. https://cirrus-ci.com/task/6161948470673408

**Comment by feeley** [comment]
If we run the CI locally we would be using machines where a `make -j` is not a problem and is fast. Here are some build times on some of these machines (for a build including the `git clone`):

```
380 seconds    gambit.iro.umontreal.ca     recent mac mini with 6 cores
240 seconds    arctic.iro.umontreal.ca     liquid cooled Debian linux PC
```

That's a lot better than the ~1200 seconds with the Kubernetes and similar to the max CPU and RAM case (with no guarantees I assume).

For the build issue, could you try removing the `-j`? I suspect it is a concurrency issue with the makefiles (probably related to git).

**Comment by feeley** [comment]
> Without `-j` the `make` succeeds in Cirrus. But one unit test fails:
> 
> ```
> [223|  0]  94% ###############. 110.9s 13-modules/prim_port.scm
> *** FAILED 13-modules/prim_port.scm WITH EXIT CODE HI=1 LO=0
> (call-with-output-process "echo" ##newline) in (include "~~lib/_prim-port#.scm") returned #!void but expected #<os-exception #2>
> (with-output-to-process "echo" ##newline) in (include "~~lib/_prim-port#.scm") returned #<os-exception #3> but expected #<os-exception #4>
> (call-with-output-process "echo" ##newline) in (namespace ("##")) returned #<os-exception #5> but expected #<os-exception #2>
> (with-output-to-process "echo" ##newline) in (namespace ("##")) returned #<os-exception #6> but expected #<os-exception #4>
> ```

I suspect a problem with subprocesses... Is there an easy way to run tests on the Cirrus machine without having to rebuild from scratch?  It really makes experimenting painful...

Anyway, please run this test:

```
;; File: process-issue.scm

(define (run n)
  (let loop ((i 0))
    (if (< i n)
        (begin
          (if (= 0 (modulo i 100)) (println "--------- " i))
          (with-exception-catcher
           (lambda (exc)
             (println "*** after " i " iterations got this exception:")
             (display-exception exc))
           (lambda ()
             (call-with-output-process "echo" ##newline)))
          (loop (+ i 1))))))

(run 10000)
```

**Comment by lassik** [comment]
> Is there an easy way to run tests on the Cirrus machine without having to rebuild from scratch? It really makes experimenting painful...

Nope, none of the CI services have ever had that. Since it's just a Docker container, you can run the base image locally, but of course it's not exactly the same environment.

**Comment by feeley** [comment]
Actually, on my macOS machine I got this:

```
% gsi process-issue.scm 
--------- 0
--------- 100
--------- 200
--------- 300
--------- 400
*** after 493 iterations got this exception:
Broken pipe
(force-output '#<input-output-port #2 (process "echo")>)
```

So this probably explains why the test is failing on Cirrus (maybe the probability of failure is higher because the `echo` process terminates faster there than on my test machine and the race is more often won by the `echo` process).

I think this should be declared a problem with the unit test.  Otherwise, if Gambit's runtime system is modified to automatically silence the broken pipe error, there might be problematic situations where you actually care to know that the process died "prematurely" and you can't know.

**Comment by lassik** [comment]
Wow, that's weird.



> I think this should be declared a problem with the unit test. Otherwise, if Gambit's runtime system is modified to automatically silence the broken pipe error, there might be problematic situations where you actually care to know that the process died "prematurely" and you can't know.

Agreed. Silencing errors is generally a recipe for worse problems later on.

I can disable that unit test and add a comment explaining why.

**Comment by feeley** [comment]
Instead of calling `echo`, call `sort` which also exists on unix and windows... The sort process will wait for the input to be provided.

**Comment by lassik** [comment]
Please check out the latest build: https://cirrus-ci.com/task/5067557282775040

The `build` stage runs fine, but your `process-issue.scm` test gives tons of `broken pipe` errors with the `echo` command all the time. So that's what caused it for sure.

The end of the  `build` stage says `**** to compile the builtin modules you should: make modules`. Is there a principled way to deal with module building in CI so that it would always work right, or are there going to be some bootstrap issues sometimes no matter how it is done?

**Comment by feeley** [comment]
Indeed the `echo` dying fast(er) on the Cirrus machine is clearly the issue.

The modules aren't built automatically by the build steps because a `make modules` builds them as dynamically loadable .o1 files (OS "shared libraries") and, although this will work fine on linux, macOS and Windows, it is possible some OS in the wild does not support this and a `make` would end with a build failure.  Given that currently Gambit can work fine without building the modules it is an optional but recommended step (none of the modules currently call FFI code or use external OS libraries).  For the CI builds a `make modules` should always work except for the setting where `--enable-ansi-c` is configured... which precludes calling `dlopen`.

**Comment by lassik** [comment]
I added a `make modules` step to the CI. In the `.yml` file we always control which exact OS runs each build, and we know that the architecture is x86-64, so it shouldn't be a problem even to have very specific settings in there. The build commands can also vary by OS.

I'll try adding MacOS and FreeBSD next. Windows is also supported, but we have to figure out how to set up the compilers and run them. I've done Cygwin and Msys gcc builds on Cirrus. MSVC must be available too but I don't know how to run it.

The make without `-j` is really slow; I'll try adding it back.

**Comment by lassik** [comment]
There appears to be something about `make -j` that causes the build to try to do things more than once and then fail. https://cirrus-ci.com/task/4792435783237632 is an example.

**Comment by lassik** [comment]
This is not supposed to happen in normal CI, right?

```
**** checking if current gsc-boot must be built
**** building boot/gsc-boot
```

**Comment by feeley** [comment]
I suspect that the `deselect-gen-for-commit` is done too early... in other words there is a missing makefile dependency for `deselect-gen-for-commit` to cause it to happen after the files are generated.

**Comment by feeley** [comment]
> This is not supposed to happen in normal CI, right?
> 
> ```
> **** checking if current gsc-boot must be built
> **** building boot/gsc-boot
> ```

Yes, but only once.  It is because a full bootstrap must happen since we are starting with a fresh clone.

**Comment by lassik** [comment]
Good to know.

MacOS CI now works, except that 2 tests fail:

```
[161|  0]  67% ##########...... 6.6s 06-thread/thread_join.scm
*** FAILED 06-thread/thread_join.scm WITH EXIT CODE HI=1 LO=0
"unit-tests/06-thread/thread_join.scm"@30.1: FAILED (check-tail-exn join-timeout-exception? (lambda () (thread-join! t2 .001))) GOT 2
"unit-tests/06-thread/thread_join.scm"@32.1: FAILED (check-equal? (thread-join! t2 -1 123) 123) GOT 2
"unit-tests/06-thread/thread_join.scm"@33.1: FAILED (check-equal? (thread-join! t2 .001 123) 123) GOT 2

[165|  1]  70% ###########..... 7.6s 06-thread/thread_start.scm
*** FAILED 06-thread/thread_start.scm WITH EXIT CODE HI=1 LO=0
"unit-tests/06-thread/thread_start.scm"@32.1: FAILED (check-equal? var 1) GOT 0
```

**Comment by feeley** [comment]
Those are probably race conditions too...

**Comment by feeley** [comment]
That is strange... the behaviour makes me suspect that the heartbeat interrupts (for preemptime scheduling of the green threads) are somehow disabled. But the previous tests indicate that the heartbeat interrupts are at ~1000HZ.  So I don't understand how that could be the case when the tests run fine on my macOS development machine (using clang and gcc-8).

**Comment by lassik** [comment]
On the bright side, FreeBSD passed easily!

**Comment by lassik** [comment]
Note that since FreeBSD and MacOS do not work inside Docker containers, they use some different kind of virtualization. But those test cases do not fail on FreeBSD. Maybe it's intermittent failure? We've only had one macOS test run so far.

**Comment by lassik** [comment]
There's now a Cygwin build: https://github.com/udem-dlteam/gambit/blob/lassik/cirrus/.cirrus.yml

Do you know why it's trying to run autoconf when the `make` command is issued? https://cirrus-ci.com/task/5241569640185856

**Comment by feeley** [comment]
Nice!  The issue may be related to incorrect modification dates on the files?  Could you `ls -l config*` just after cloning?

**Comment by lassik** [comment]
```
-rwx------+ 1 ContainerAdministrator ContainerAdministrator  44166 Nov  6 14:09 config.guess
-rwx------+ 1 ContainerAdministrator ContainerAdministrator  36141 Nov  6 14:09 config.sub
-rwx------+ 1 ContainerAdministrator ContainerAdministrator 455587 Nov  6 14:09 configure
-rwx------+ 1 ContainerAdministrator ContainerAdministrator 114207 Nov  6 14:09 configure.ac
```

**Comment by feeley** [comment]
Maybe the issue shows up because of the coarse time stamp granularity on Windows, which is 2 seconds I think.  So perhaps `make` can't tell that `configure` is newer than `configure.ac`.  Could you add to the CI script a `touch configure` to force a newer time stamp?

**Comment by lassik** [comment]
It's only 2 seconds on the old FAT file systems; NTFS timestamps use "100-nanosecond intervals".

**Comment by feeley** [comment]
OK... but can you still try adding a touch?  If that doesn't work then it might be a deeper problems with the makefiles or cygwin's make.

**Comment by lassik** [comment]
MacOS passed in the latest build. That means the failures in the thread tests are intermittent.

**Comment by lassik** [comment]
Since this issue is related to core Gambit instead of libs, can you [move it](https://help.github.com/en/github/managing-your-work-on-github/transferring-an-issue-to-another-repository) into gambit/gambit?

**Comment by lassik** [comment]
`touch configure` fixed the issue of Cygwin trying to run autoconf.

**Comment by lassik** [comment]
Also on Cygwin:

```
checking ws2tcpip.h usability... no
checking ws2tcpip.h presence... yes
configure: WARNING: ws2tcpip.h: present but cannot be compiled
configure: WARNING: ws2tcpip.h:     check for missing prerequisite headers?
configure: WARNING: ws2tcpip.h: see the Autoconf documentation
configure: WARNING: ws2tcpip.h:     section "Present But Cannot Be Compiled"
configure: WARNING: ws2tcpip.h: proceeding with the compiler's result
configure: WARNING:     ## -------------------------------------- ##
configure: WARNING:     ## Report this to gambit@iro.umontreal.ca ##
configure: WARNING:     ## -------------------------------------- ##
checking for ws2tcpip.h... no
```

**Comment by lassik** [comment]
All platforms green! https://cirrus-ci.com/build/6149589735833600

**Comment by feeley** [comment]
Wonderful! So Cirrus appears to suit our needs.

Can you also add builds with mingw and Microsoft compilers?  And also build status feedback for the README badges? With Travis and Appveyor I also was building with C++ (to make sure Gambit builds with either a C or C++ compiler), and also with `--enable-multiple-threaded-vms` (to test the parallel GC), and also with `--enable-ansi-c` (to test a bare-bones build).

When all of this is supported I'll accept a pull request to switch Gambit to Cirrus CI.

Separately, and likely after switching to Cirrus CI, we can add some settings such as the universal backend and the native backend.  That will be many settings to test!  But it will be a good way to avoid bitrot with those backends.

**Comment by lassik** [comment]
> Can you also add builds with mingw and Microsoft compilers?

We can do that. Gambit's current AppVeyor config has:

```
    - arch: "x86-64"
      tools: "mingw"

    - arch: "x86-64"
      tools: "msvc"

    - arch: "x86-64"
      tools: "vstudio.bat"

    - arch: "x86-64"
      tools: "cygwin"

    - arch: "x86-32"
      tools: "msvc"

    - arch: "x86-32"
      tools: "vstudio.bat"
```

Of these, Cirrus currently covers just 64-bit Cygwin. I wonder what are the easiest ways to install the requisite toolchains for the other builds. The Windows image provided by Cirrus comes pre-installed with the Chocolatey package manager (`choco` command) so that may be the easiest option. Scoop will probably be easy to install if needed. I'm not sure how to install MSVC; it may be pre-installed.

Chololatey has at least one MingW package: https://chocolatey.org/packages/mingw

**Comment by lassik** [comment]
The very portable libarchive is using Cirrus, so we can adapt what they do: https://github.com/libarchive/libarchive/tree/master/build/ci/cirrus_ci

**Comment by feeley** [comment]
Yes that build script for the choice of Windows dev tools seems well written.  My only comment would be to use the ".bat" extension instead of ".cmd" (".bat" is used in many places in Gambit and it is easier to grep if only one extension is used).

**Comment by lassik** [comment]
The MSVC compiler is `cl` but what do we use for `make` - should we install GNU make from MingW, or does Visual Studio ship with a version of `make` that works with Gambit's autotools?

**Comment by lassik** [comment]
Would it make sense to run `make install` as part of CI builds, and roll the installation directory into a tar file that is shipped as a build artifact? That would be a really easy way for people to install and test cutting-edge builds of Gambit on popular platforms without compiling anything. It would also be an easy way to get a `gsc-boot` binary for compiling the master branch.

**Comment by lassik** [comment]
We could do something like `./configure --prefix=/usr/local/Gambit/$branch` (as discussed earlier in this issue) and then `tar -czf gambit-$branch-$commithash.tgz` and tell Cirrus to preserve that tar file.

**Comment by feeley** [comment]
The only problem is that the directory in which the tar file is untared must be known when the CI build is done. The DESTDIR environment variable could be used if the install path on the CI machine and the target machine don't match.

**Comment by lassik** [comment]
Yeah, users would have to use the same directory locally. `$HOME/.local/Gambit/...` would be a good location but since `$HOME` varies per user, I assume it doesn't work.

The trouble with the CI artifacts is we don't know anything about the target machine since any number of users might download the tar files onto different machines/user accounts. We'd to agree on a convention.

**Comment by feeley** [comment]
That would be useful, but it should be seen as a "fallback" and not be part of the standard way of building Gambit (otherwise building Gambit will depend on CI, which is kind of backwards).

**Comment by feeley** [comment]
I don't recall, but maybe the installation directory can be relative to the user's home (if it contains a ~/ at the start).

**Comment by feeley** [comment]
That doesn't work because the configure script insists on an absolute path.

**Comment by lassik** [comment]
Agreed. I meant the CI tarballs as an easy alternative to building from source, not as a replacement. Some projects (operating systems, etc.) that are complex to buil publish frequent binary snapshots. People who want to follow cutting-edge stuff but don't need the up-to-the-minute latest source code can use those, sparing the effort of compiling it themselves. I ran OpenBSD snapshots back in the day, as did many other people, and it was a good experience. Nowadays CI is ubiquitous so it's simply the most natural place to generate those snapshots (especially as CI runs tests, so it can automatically publish a new snapshot whenever tests pass).

If `~` in the prefix works, that would probably be easiest. But since you are GNU autotools compatible, I suspect it doesn't.

Gambit has the `-:~~` runtime option to override installation directories; would that work?

**Comment by feeley** [comment]
Yes `-:~~=...` would work but will be painful.  A slightly better option is to set `GAMBOPT=~~=...`.  The best would be to have `~/.gambit/<hash>` or whatever as a `GAMBITDIR`.  Will check if this is possible.

**Comment by lassik** [comment]
The experience of following the snapshots was considerably different from running the latest release versions, as they constantly added new features/bugfixes and kept their main development branch quite stable so it was very rare for any features to be broken. This is the situation for Gambit as well. Would be interesting to find out how much demand there is for this thing :) Best case, it would attract new contributors to file bugs / PRs.

**Comment by lassik** [comment]
`GAMBITDIR` exists already in the repo but is not mentioned in the manual. Is it accessible to users yet?

**Comment by feeley** [comment]
`GAMBITDIR` is a build time symbol, not an environment variable.  It is set by the configure script to the same value as `prefix`.

By the way, it seems to work when `prefix` is forced to `~/.gambit/foobar` by adding an explicit assignment in the configure script.  After a `make`, I get:

```
> (path-expand "~~")
"/Users/feeley/.gambit/foobar/"
```

But the `make install` failed to actually install in `~/.gambit/foobar` for an unknown reason.  But maybe adding the `DESTDIR` would solve that. I'll let you experiment... gtg...

**Comment by lassik** [comment]
Hmm. Did you manually change `./configure` in a text editor? It's awesome that you got it to work that far but I'd start from the assumption that Autotools is "Jenga code" (https://en.wikipedia.org/wiki/Jenga) - one false move and everything falls down :)

**Comment by lassik** [comment]
Cirrus isn't picking up commits from the recreated udem-dlteam/gambit

**Comment by feeley** [comment]
Any progress on this front?  It will be good to have a central CI system to build upon as libraries will also need to use CI, in particular the `gambit/ffi/python` library that is advancing well but needs tests on linux/macOS/Windows to verify that it works on the main platforms.

**Comment by feeley** [comment]
I just learned that Travis-CI now supports all 3 platforms, and that github has their own CI.  Is there a good reason to consider Cirrus CI?

**Comment by daviwil** [comment]
Just an outsider with an opinion here :)

I've had a great experience using GitHub Actions to automate a cross-platform Gambit build for my own use:

https://github.com/daviwil/gambit/blob/build-gambit/.github/workflows/build-gambit.yml

If anyone is interested in using Actions I'd be happy to help answer questions or set something up. Nice thing about Actions is that it's easy to create a custom task that can be used across multiple repos, probably something that would be useful if Gambit's modules will be split into their own repositories.

**Comment by feeley** [comment]
Yes Actions seems like a good choice given it covers all majors platforms.

Could you look into replacing the current Travis-CI and Appveyor scripts?

**Comment by lassik** [comment]
@daviwil If you already have a solid understanding of Actions, trying it out sounds good.

**Comment by daviwil** [comment]
Sure!  I'll put together a draft CI config that replaces what we've got for TravisCI and AppVeyor and send along a PR in a couple days.

**Comment by feeley** [comment]
Thanks!

**Comment by lassik** [comment]
Closing this issue since @daviwil has done good work on GitHub Actions based CI and there are no impediments in sight to adopting it wholesale for the main Gambit repo.
